llama-index-experimental/llama_index/experimental/retrievers/llama-index-retrievers-natural-language/nl_data_frame_retierver.py

```diff
@@ -0,0 +1,182 @@
+from llama_index.core.callbacks.base import CallbackManager
+from llama_index.core.constants import DEFAULT_SIMILARITY_TOP_K
+from llama_index.core.llms import llm
+from llama_index.core.llms.utils import resolve_llm
+from llama_index.core.utils import print_text
+from llama_index.core.retrievers import BaseRetriever
+from llama_index.core.schema import NodeWithScore, QueryBundle, TextNode
+from llama_index.core.prompts import BasePromptTemplate
+from llama_index.core.prompts.default_prompts import DEFAULT_TEXT_TO_SQL_PROMPT
+from typing import List, Optional
+
+import asyncio
+import logging
+import pandas as pd
+import duckdb
+import re
+
+logger = logging.getLogger(__name__)
+
+
+class NLDataframeRetriever(BaseRetriever):
+    def __init__(
+        self,
+        df: pd.DataFrame,
+        llm: llm,
+        name: Optional[str] = None,
+        text_to_sql_prompt: Optional[BasePromptTemplate] = None,
+        similarity_top_k: int = DEFAULT_SIMILARITY_TOP_K,
+        callback_manager: Optional[CallbackManager] = None,
+        verbose: bool = False,
+    ):
+        self._llm = resolve_llm(llm)
+        self._similarity_top_k = similarity_top_k
+        self._text_to_sql_prompt = text_to_sql_prompt or DEFAULT_TEXT_TO_SQL_PROMPT
+        data_source = df.copy(deep=True)
+        data_source.rename(columns=lambda x: re.sub(r"\s+", "_", x), inplace=True)
+        table_name = name or "data_table"
+        self._connection = duckdb.connect()
+        self._connection.sql(f"CREATE TABLE {table_name} AS SELECT * FROM data_source")
+        self._connection.sql(
+            f"INSERT INTO {table_name} BY NAME SELECT * FROM data_source"
+        )
+        self._schema = self._connection.table(f"{table_name}").describe()
+        self._description = None
+        self._owl = None
+        self._schema_str = self._create_schema()
+
+        super().__init__(callback_manager=callback_manager, verbose=verbose)
+
+    def get_owl(self) -> str:
+        if self._owl is None:
+            response = self._llm.complete(
+                f"""
+    given the schema:{self._schema_str}
```

Comment:
We might want to make prompts customizable, but up to you

Reply:
Makes total sense and will do those changes!